std/array.d

```diff
@@ -3927,3 +3927,213 @@ unittest
     assert(appS.data == "hellow");
     assert(appA.data == "hellow");
 }
+
+/++
+Params: a = The input elements
+
+Returns: A static array constructed from `a`.
+
+The type of elements can be specified implicitly (`int[2] a = [1,2].asStatic;`)
+or explicitly (`float[2] a = [1,2].asStatic!float`). When `a` is a range,
+the number of elements has to be given as template argument (eg `2.iota.asStatic!2`).
+Size and type can be combined (eg: `2.iota.asStatic!(byte[2])`).
+Range `a` can also be specified as a template argument (eg: `asStatic!(2.iota)`
+or `asStatic!(double, 2.iota)`).
+
+Note: `foo([1, 2, 3].asStatic)` may be inefficient because of the copies involved.
++/
+pragma(inline, true) T[n] asStatic(T, size_t n)(auto ref T[n] a) nothrow @safe pure @nogc
+{
+    return a;
+}
+
+/// ditto
+U[n] asStatic(U, T, size_t n)(auto ref T[n] a) nothrow @safe pure @nogc
```

Comment:
Explicitly specifying these attributes has the downside of erroring if a user has an unsafe destructor. The compiler already infers the attributes for you and you check this with your unittests, so what's the motivation for being explicit here?

Reply:
done, note that this rule isn't always observed eg:
```
auto uninitializedArray(T, I...)(I sizes) nothrow @system
std/complex.d:37:24:auto complex(R)(R re)  @safe pure nothrow @nogc
pure bool isLoggingEnabled()(LogLevel ll) @safe nothrow @nogc
std/math.d:646:23:auto conj(Num)(Num z) @safe pure nothrow @nogc
```

Reply:
Well not all phobos code is created equal. There's a lot of old, ugly legacy code in Phobos (though it's not as bad as DMD).
Anyhow, fixes for the lines you observed are always welcome.

Reply:
ack

Reply:
=> filed https://issues.dlang.org/show_bug.cgi?id=18618 Issue 18618 - templated functions should in general have their attributes inferred (edit)